[WebGL] Only transfer required bytes over IPC for texture initialisation

djg opened the pull request:
#### 1104571ccd3419d753fa3cb57f8f04300466a9f3
<pre>
[WebGL] Only transfer required bytes over IPC for texture initialisation
<a href="https://bugs.webkit.org/show_bug.cgi?id=242614">https://bugs.webkit.org/show_bug.cgi?id=242614</a>
rdar://94539892

Reviewed by Dean Jackson.

With &quot;GPUP: WebGL&quot; enabled, WebGL texture initialisation functions, texImage2D,
et al., are passing too much data to the GPUP via IPC. The extra data transfer
causes a performance regression in the reported webpage.

Introduce a helper layer that calculates the exact number of bytes required by
texture initialisation functions and use this value when creating GCGLSpan
instead.

* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::texImageSourceHelper):
Use helper functions
(WebCore::WebGLRenderingContextBase::computeImageSizeInBytes):
Helper to compute the number of bytes required by a texture initialisation function.
(WebCore::WebGLRenderingContextBase::texImageArrayBufferViewHelper):
(WebCore::WebGLRenderingContextBase::texImageImpl):
Use helper functions
(WebCore::WebGLRenderingContextBase::texImage2DBase):
(WebCore::WebGLRenderingContextBase::texImage3DBase):
(WebCore::WebGLRenderingContextBase::texSubImage2DBase):
(WebCore::WebGLRenderingContextBase::texSubImage3DBase):
Calculate the required bytes and use the minimum of require bytes and buffer
length when calling into GL.
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:

Canonical link: <a href="https://commits.webkit.org/252373@main">https://commits.webkit.org/252373@main</a>
</pre>

grorg commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -5339,6 +5356,12 @@ void WebGLRenderingContextBase::texImageImpl(TexImageFunctionID functionID, GCGL
+    bool ok = computeImageSizeInBytes(TexImageDimension::Tex2D, format, type, width, height, 1, &bytesRequired);
+    ASSERT_UNUSED(ok, ok, "Parameters should already have been validated!");
+    // FIXME: ANGLE robust API uses signed int for byte count.
+    ASSERT(isInBounds<GCGLsizei>(bytesRequired));
```
Should we throw an error if our `unsigned` value is out of the signed bounds?

djg committed:
[WebGL] Only transfer required bytes over IPC for texture initialisation
https://bugs.webkit.org/show_bug.cgi?id=242614
rdar://94539892

Reviewed by Dean Jackson.

With "GPUP: WebGL" enabled, WebGL texture initialisation functions, texImage2D,
et al., are passing too much data to the GPUP via IPC. The extra data transfer
causes a performance regression in the reported webpage.

Introduce a helper layer that calculates the exact number of bytes required by
texture initialisation functions and use this value when creating GCGLSpan
instead.

* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::texImageSourceHelper):
Use helper functions
(WebCore::WebGLRenderingContextBase::computeImageSizeInBytes):
Helper to compute the number of bytes required by a texture initialisation function.
(WebCore::WebGLRenderingContextBase::texImageArrayBufferViewHelper):
(WebCore::WebGLRenderingContextBase::texImageImpl):
Use helper functions
(WebCore::WebGLRenderingContextBase::texImage2DBase):
(WebCore::WebGLRenderingContextBase::texImage3DBase):
(WebCore::WebGLRenderingContextBase::texSubImage2DBase):
(WebCore::WebGLRenderingContextBase::texSubImage3DBase):
Calculate the required bytes and use the minimum of require bytes and buffer
length when calling into GL.
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:

Canonical link: https://commits.webkit.org/252373@main
